qatzip/compress.go

```diff
@@ -340,20 +342,24 @@ func (z *Writer) compressWrite(p []byte) (n int, err error) {
 		z.traceLogf(Med, "[write->qat] r:%v i:%v o:%v ibofs:%v obl:%v err:%v", remainder, in, out, consumed, z.outputBuf.Len(), err)
 
 		if err != nil {
-			if err == ErrBuffer {
-				// expand output buffer
-				// TODO grow to a maximum size
+			if err == ErrBuffer && !isPartialJob {
+				// Grow output buffer, only in the case that it was not large enough for QAT to output any data.
+				// If QAT does output data, we save a copy by doing it after the output buffer draining.
+
 				t1 = time.Now().UnixNano()
-				z.bufferGrowth *= 2
-				newSize := remainder + z.bufferGrowth
-				z.traceLogf(Med, "[expand output buffer] o:%v n:%v", z.outputBuf.Len(), newSize)
-				z.outputBuf = bytes.NewBuffer(make([]byte, newSize))
+				newLength := z.outputBuf.Len()*z.bufferGrowth
```

Comment:
I mentioned below that we should not change the API. I think we can get away with just adding buffer growth value to the size of output buffer. There is an edge case we need to test for when the input size is 2GB and it has no way to expand it could cause a infinite loop. You need to just return an error in that case.